src/sciline/reporter.py

```diff
@@ -0,0 +1,240 @@
+    """Base class for progress reporters of computations."""
+
+    def __init__(self) -> None:
+        self._n_steps = 0
+        self._provider_id = -1
+
+    def run_computation(self, providers: Iterable[Provider]) -> Reporter:
+        """Run a single computation with progress reporting.
+
+        A ``Reporter`` instance must not be used for multiple computations concurrently!
+
+        Use this method in a ``with`` statement:
+
+        .. code-block:: python
+
+            with reporter.run_computation(providers):
+                # Perform computation
+
+        This method exists mainly to provide arguments when entering a reporting
+        context, implementers should mainly override ``__enter__``.
+        """
+        self._n_steps = sum(1 for provider in providers if provider.kind == "function")
+        return self
+
+    @abstractmethod
+    def __enter__(self) -> None:
+        """Computation is starting."""
+
+    @abstractmethod
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        """Computation is finished."""
+
+    @abstractmethod
+    def on_provider_start(self, provider: Provider) -> int:
+        """Called when a provider is started."""
+
+    @abstractmethod
+    def on_provider_end(self, provider_id: int) -> None:
+        """Called when a provider has finished."""
+
+    def reporting_provider_func(self, provider: Provider) -> Callable[..., Any]:
+        """Wrap a provider's function to report progress with this reporter."""
+        if provider.kind != "function":
+            return provider.func
+
+        @functools.wraps(provider.func)
+        def reporting_func(*args: Any, **kwargs: Any) -> Any:
+            provider_id = self.on_provider_start(provider)
+            try:
+                result = provider.func(*args, **kwargs)
+            finally:
+                self.on_provider_end(provider_id)
+            return result
+
+        return reporting_func
+
+    def call_provider_with_reporting(
+        self, provider: Provider, values: dict[Hashable, Any]
+    ) -> Any:
+        """Call a provider and report its progress with this reporter."""
+        if provider.kind != "function":
+            return provider.call(values)
+
+        provider_id = self.on_provider_start(provider)
+        try:
+            result = provider.call(values)
+        finally:
+            self.on_provider_end(provider_id)
+        return result
+
+    def _get_provider_id(self) -> int:
+        """Return a unique identifier for a current provider."""
+        self._provider_id += 1
+        return self._provider_id
+
+
+class RichReporter(Reporter):
+    """Report progress using Rich.
+
+    This class uses the `rich <https://rich.readthedocs.io/en/stable/index.html>`_
+    package to display a progress bar.
+    Rich is not a hard dependency of Sciline and must be installed separately.
```

Comment:
We should add `rich` to one of the optional deps group (or create a new one). It's easier to keep a track of deps in one single location (pyproject.toml).